stl/inc/ratio

```diff
@@ -19,59 +20,58 @@ _STL_DISABLE_CLANG_WARNINGS
 #undef new
 
 _STD_BEGIN
-template <intmax_t _Val>
-struct _Abs : integral_constant<intmax_t, (_Val < 0 ? -_Val : _Val)> {}; // computes absolute value of _Val
-
-template <intmax_t _Ax, intmax_t _Bx, bool _Sfinae = false,
-    bool _Good = (_Abs<_Ax>::value <= INTMAX_MAX / (_Bx == 0 ? 1 : _Abs<_Bx>::value))>
-struct _Safe_mult : integral_constant<intmax_t, _Ax * _Bx> {}; // computes _Ax * _Bx without overflow
+_NODISCARD constexpr intmax_t _Abs(const intmax_t _Val) noexcept { // computes absolute value sof _Val
+    return _Val < 0 ? -_Val : _Val; // caller ensures that _Val does not go below -INTMAX_MAX
+}
 
-template <intmax_t _Ax, intmax_t _Bx, bool _Sfinae>
-struct _Safe_mult<_Ax, _Bx, _Sfinae, false> { // _Ax * _Bx would overflow
-    static_assert(_Sfinae, "integer arithmetic overflow");
-};
+_NODISCARD constexpr intmax_t _Safe_mult(
+    const intmax_t _Ax, const intmax_t _Bx) noexcept { // computes _Ax * _Bx without overflow
+    if (_Abs(_Ax) > INTMAX_MAX / (_Bx == 0 ? 1 : _Abs(_Bx))) {
+        _CSTD abort(); // assert does not work in NDEBUG
+    }
 
-template <intmax_t _Val>
-struct _Sign_of : integral_constant<intmax_t, (_Val < 0 ? -1 : 1)> {}; // computes sign of _Val
+    return _Ax * _Bx;
+}
 
-template <intmax_t _Ax, intmax_t _Bx, bool _Good, bool _Also_good>
-struct _Safe_addX : integral_constant<intmax_t, _Ax + _Bx> {}; // computes _Ax + _Bx without overflow
+_NODISCARD constexpr intmax_t _Sign_of(const intmax_t _Val) noexcept { // computes sign of _Val
+    return _Val < 0 ? -1 : 1;
+}
 
-template <intmax_t _Ax, intmax_t _Bx>
-struct _Safe_addX<_Ax, _Bx, false, false> { // _Ax + _Bx would overflow
-    static_assert(_Always_false<_Safe_addX>, "integer arithmetic overflow");
-};
+_NODISCARD constexpr intmax_t _Safe_add(
+    const intmax_t _Ax, const intmax_t _Bx) noexcept { // computes _Ax + _Bx without overflow
+    if (_Sign_of(_Ax) == _Sign_of(_Bx) && (_Abs(_Ax) > INTMAX_MAX - _Abs(_Bx))) {
+        _CSTD abort(); // assert does not work in NDEBUG
+    }
 
-template <intmax_t _Ax, intmax_t _Bx>
-struct _Safe_add : _Safe_addX<_Ax, _Bx, _Sign_of<_Ax>::value != _Sign_of<_Bx>::value,
-                       (_Abs<_Ax>::value <= INTMAX_MAX - _Abs<_Bx>::value)>::type {
-    // computes _Ax + _Bx, forbids overflow
-};
+    return _Ax + _Bx;
+}
 
-template <intmax_t _Ax, intmax_t _Bx>
-struct _GcdX : _GcdX<_Bx, _Ax % _Bx>::type {}; // computes GCD of _Ax and _Bx
+_NODISCARD constexpr intmax_t _Gcd(
+    const intmax_t _Ax, const intmax_t _Bx) noexcept { // computes GCD of abs(_Ax) and abs(_Bx)
+    if (_Ax == 0 && _Bx == 0) {
+        return 1; // contrary to mathematical convention; avoids division by 0 in ratio_less
+    }
 
-template <intmax_t _Ax>
-struct _GcdX<_Ax, 0> : integral_constant<intmax_t, _Ax> {}; // computes GCD of _Ax and 0
+    intmax_t _Temp_Ax = _Abs(_Ax);
+    intmax_t _Temp_Bx = _Abs(_Bx);

```

Comment:
Actually there's no rule that you must *always* use `const` params, you just mark `const` where they happen to be `const`. Here they can be non-`const` to avoid extra variables.

Reply:
Applied